qiskit_ode/signals/signals.py

```diff
@@ -27,167 +28,196 @@
+        self._phase_arg = None
 
-    @abstractmethod
-    def value(self, t: float = 0.0) -> float:
-        """Return the value of the signal at time t."""
+        # set carrier and phase
+        self.carrier_freq = carrier_freq
+        self.phase = phase
 
     @property
     def name(self) -> str:
         """Return the name of the signal."""
         return self._name
 
-    def to_pwc(self, dt: float, n_samples: int, start_time: float = 0.0) -> "PiecewiseConstant":
-        """
-        Converts a signal to a piecewise constant signal.
+    @property
+    def carrier_freq(self) -> Array:
+        """The carrier frequency of the signal."""
+        return self._carrier_freq
 
-        Args:
-            dt: Time increment to use.
-            n_samples: number of steps to resample with.
-            start_time: start time from which to resample.
+    @carrier_freq.setter
+    def carrier_freq(self, carrier_freq: float):
+        """Carrier frequency setter."""
+        if type(carrier_freq) == list:
+            carrier_freq = [Array(entry).data for entry in carrier_freq]
+        self._carrier_freq = Array(carrier_freq)
+        self._carrier_arg = 1j * 2 * np.pi * self._carrier_freq
 
-        Returns:
-            A piecewiseConstant signal.
-        """
-        samples = [self.envelope_value(dt * idx + start_time) for idx in range(n_samples)]
+    @property
+    def phase(self) -> Array:
+        """The phase of the signal."""
+        return self._phase
 
-        return PiecewiseConstant(
-            dt, samples, start_time=start_time, carrier_freq=self.carrier_freq, phase=self.phase
-        )
+    @phase.setter
+    def phase(self, phase: float):
+        """Phase setter."""
+        if type(phase) == list:
+            phase = [Array(entry).data for entry in phase]
+        self._phase = Array(phase)
+        self._phase_arg = 1j * self._phase
 
-    def __mul__(self, other):
-        return signal_multiply(self, other)
+    def envelope(self, t: Union[float, np.array, Array]) -> Union[complex, np.array, Array]:
+        """Vectorized evaluation of the envelope at time t."""
+        return self._envelope(t)
 
-    def __rmul__(self, other):
-        return signal_multiply(self, other)
+    def complex_value(self, t: Union[float, np.array, Array]) -> Union[complex, np.array, Array]:
+        """Vectorized evaluation of the complex value at time t."""
+        arg = self._carrier_arg * t + self._phase_arg
+        return self.envelope(t) * np.exp(arg)
 
-    def __add__(self, other):
-        return signal_add(self, other)
+    def __call__(self, t: Union[float, np.array, Array]) -> Array:
+        """Vectorized evaluation of the signal at time t."""
+        return np.real(self.complex_value(t))
 
-    def __radd__(self, other):
-        return signal_add(self, other)
+    def __str__(self) -> str:
+        """Return string representation."""
+        if self.name is not None:
+            return str(self.name)
```

Comment:
Would it perhaps not be nicer to return
```
Signal(carrier_freq={freq}, phase={phase}, name={name})
```
if it has a name?

Reply:
My thinking was that if it has a name it should just be displayed to make it more succinct, but I could go either way on this.

Reply:
Ok. We can leave as is for now. It is easy to change later on and will not have a big impact.